File: python/50_consecutive_prime_sum.py
def consecutive_prime_sum() -> int:
    primes = []
    sum: int = 0
    max_terms: int = 0
    
    sieved = [False] * 1_000_000
    for i in range(2, 1_000_000):
        if not sieved[i]:
            for j in range (2*i, 1_000_000, i):
                sieved[j] = True
            primes.append(i)
    primes_set = set(primes)
            
    for prime in primes:
        if sum + prime < 1_000_000:
            sum += prime
            max_terms += 1
        else:
            break
    
    sums = [0] * (max_terms + 1)
    sums[0] = 2
    for i in range(1, max_terms + 1):
        sums[i] = sums[i - 1] + primes[i]
    
    ans = 0
    for len in range(max_terms, 0, -1):
        for start in range(max_terms - len - 1, 0, -1):
            sum_a = sums[start]
            sum_b = sums[start + len]
            
            if sum_b - sum_a in primes_set:
                ans = sum_b - sum_a
                break
        if ans != 0:
            break
    return ans

if __name__ == "__main__":
    print(consecutive_prime_sum())

File: cpp/50_consecutive_prime_sum.cpp
#include <iostream>
#include <vector>
#include <set>

int consecutive_prime_sum(){
    std::vector<int> primes;
    std::set<int> primesSet;
    bool sieved[1'000'000];
    for (int i = 2; i < 1'000'000; i++){
        if (!sieved[i]){
            for (int j = 2*i; j < 1'000'000; j += i){
                sieved[j] = true;
            }
            primes.push_back(i);
            primesSet.insert(i);
        }
    }
    // find maximum number of primes that can sum to number below 1 million
    int sum = 0, maxTerms = 0;
    for (int prime : primes){
        if (sum + prime > 1'000'000) break;
        sum += prime;
        maxTerms++;
    }
    // sums[i] is cumulative sum of first i primes, sums[0] = 0
    int sums[maxTerms + 1] = {0};
    for (int i = 1; i <= maxTerms; i++){
        sums[i] = sums[i - 1] + primes[i - 1];
    }
    // examine consecutive sum of n primes starting from (start + 1)th prime
    for (int terms = maxTerms; terms >= 0; terms--){
        for (int start = maxTerms - terms; start >= 0; start--){
            int sumA = sums[start], sumB = sums[start + terms];
            if (primesSet.find(sumB - sumA) != primesSet.end()){
                return sumB - sumA;
            }
        }
    }
    return 0;
}

int main(){
    std::cout << consecutive_prime_sum() << "\n";
}
